File: Arrays_and_Hashing/Python/238_product_of_array_except_self.py
# RUNTIME: 194 ms, faster than 60.38% of python3 Submissions. 
# MEMORY USAGE: 25.5 MB, less than 12.52% of python3 Submissions.

from __future__ import annotations
from ast import *


# My thougtprocess:
# 1. Get length of the list
# 2. initiate three lists containing 1s with equal length
# 3. interate through the list and calculate the left products and then the right side
# 4. interate through the list and calculate the result

class Solution:
    def productExceptSelf(self, nums: List[int]) -> List[int]:
        n = len(nums)
        left_products = [1] * n
        right_products = [1] * n
        result = [1] * n
        
        left_product = 1
        for i in range(n):
            left_products[i] = left_product
            left_product *= nums[i]
        
        right_product = 1
        for i in range(n - 1, -1, -1):
            right_products[i] = right_product
            right_product *= nums[i]
        
        for i in range(n):
            result[i] = left_products[i] * right_products[i]
        
        return result

File: Arrays_and_Hashing/C++/238_product_of_array_except_self.cpp
/*
* RUNTIME: 19 ms - Beats 54.04% of submissions
* MEMORY - 26.57 MB - Beats 5.25% of submissions
*/

#include <vector>
#include <string>
#include <map>
#include <algorithm>
using namespace std;


/* Thoughtprocess:
*  1. store the size of the input vector
*  2. initiate three vectors of size n: left, right and result
*  3. calculate the product of all elemeents without the first element and store it in left
*  4. calculate the product of all elemeents without the last element and store it in right
*  5. multiply the elements of left and right and store it in result
*/


class Solution {
public:
    vector<int> productExceptSelf(vector<int>& nums) {
        int n = nums.size();
        vector<int> result;
        vector<int> left(n, 1);
        vector<int> right(n, 1);

        int product = 1;
        for (int i = 0; i < n; i++)
        {
            left[i] = product;
            product *= nums[i];
        }
        product = 1;
        for (int i = n - 1; i >= 0; i--)
        {
            right[i] = product;
            product *= nums[i];
        }
        for (int i = 0; i < n; i++)
        {
            result.push_back(left[i] * right[i]);
        }
        return result;
    }
};

